WIP: BXMSDOC-4468-master: Update decision engine doc for 7.5.

sterobin opened the pull request:
@mariofusco , @baldimir, can you SME and QE verify the following added sections in the decision engine doc? I single-sourced from community, updated, reorganized a bit, etc.

New sections from community, scattered throughout in the relevant locations:
* [KIE session pools](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#kie-sessions-pools-con_decision-engine)
* [Fact equality modes](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#fact-equality-modes-con_decision-engine)
* [Rule execution modes](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#rule-execution-modes-con_decision-engine)
* [Fact propagation modes](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#fact-propagation-modes-con_decision-engine)
* [Agenda evaluation filters](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#agenda-filters-con_decision-engine)
* [Rule units in DRL rule sets](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#drl-rule-units-con_decision-engine) (plus the three sub-sections) - I also included this section in the DRL doc because I think it's relevant there too
* [Decision engine queries and live queries](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#engine-queries-con_decision-engine)
* [Decision engine event listeners and debug logging](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#engine-event-listeners-con_decision-engine)

Updated sections:
* [Rule base configuration](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#rule-base-configuration-con_decision-engine) - Added kmodule.xml config option for sequential mode
* [KIE base attributes](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_PD/#project-kbase-ref_packaging-deploying) (from packaging doc) - Ditto, added sequential mode to table

Links:
* JIRA
* DM doc preview
* PAM doc preview - Same, but for PAM
* Drools community doc preview - Same, but for community

baldimir commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/kie-sessions-pools-con.adoc:
```diff
@@ -0,0 +1,26 @@
+[source,java]
+----
+// Obtain a KIE session pool from the KIE container
+KieContainerSessionsPool KieContainer.newKieSessionsPool(int 10)
```
This is probably a copy paste error. This whole example should look like: 

```
// Obtain a KIE session pool from the KIE container
KieContainerSessionsPool pool = kContainer.newKieSessionsPool(10);
// Create KIE sessions from the KIE session pool
KieSession kSession = pool.newKieSession();
```

sterobin replied:
I see. The community content showed it a little differently. My bad. Fixed.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/fact-equality-modes-con.adoc:
```diff
@@ -0,0 +1,54 @@
+Person p2 = new Person("John", 45);
+----
+
+In `identity` mode, facts `p1` and `p2` are treated as separate objects because they have separate identities. In `equality` mode, facts `p1` and `p2` are treated as the same object because they are composed the same way. This difference in behavior affects how you can interact with fact handles.
```
Agreed, fixed.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/fact-equality-modes-con.adoc:
```diff
@@ -0,0 +1,54 @@
+
+In `identity` mode, facts `p1` and `p2` are treated as separate objects because they have separate identities. In `equality` mode, facts `p1` and `p2` are treated as the same object because they are composed the same way. This difference in behavior affects how you can interact with fact handles.
+
+For example, assume that you insert `p1` into the {DECISION_ENGINE} and later you want to retrieve the fact handle for that fact. In `identity` mode, you must specify `p1` to return the fact handle for that exact object, whereas in `equality` mode, you can specify either `p1` or `new Person("John", 45)` to return the fact handle.
```
Ah, right. Done.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/fact-equality-modes-con.adoc:
```diff
@@ -0,0 +1,54 @@
+ksession.getFactHandle(new Person("John", 45));
+----
+
+You can set the fact equality mode using the `equalsBehavior="<mode>"` KIE base attribute in the KIE module descriptor file (`kmodule.xml`) for your {PRODUCT} project:
```
Done.

baldimir replied:
There is now "Set the system property drools.equalityBehavior to true." This is incorrect. It should be "Set the system property drools.equalityBehavior to value `equality`." 

The actual value is not true/false, but equality/identity.

sterobin replied:
Oops! Shoot. Fixed. What happened was, I use the roughly the same structure several other places and I copied/pasted for consistency and updated the content, and left that out. So glad you caught that! I changed to this:

> Set the system property `drools.equalityBehavior` to `identity` (default) or `equality`.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/rule-execution-modes-con.adoc:
```diff
@@ -0,0 +1,132 @@
+
+In this example, you can alternatively retrieve the fact handle directly from the KIE session:
+
+.Alternative application code portion to retrieve the fact handle from the KIE session
```
Sure. Was just that way in community as a reminder of what the alternative would be. But I agree. Removed.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rule-units-con.adoc:
```diff
@@ -0,0 +1,601 @@
+end
+----
+
+WARNING: Do not mix rules with and without a rule unit in the same KIE base.
```
@mariofusco Let us know what you think.

mariofusco replied:
The system doesn't allow to mix the 2 paradigms. If you try to have rules with and without unit in the same rule base you will get a compile time error.

sterobin replied:
K, I'll just change to this:

```
WARNING: Do not mix rules with and without a rule unit in the same KIE base. Mixing two rule paradigms in a KIE base results in a compilation error.
```

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rule-units-con.adoc:
```diff
@@ -0,0 +1,601 @@
+
+The example execution code produces the following output:
+
+.Example rule unit execution output
```
Done, changed to the following in both instances:

```
The example execution code produces the following output when the relevant `Person` facts are inserted in the `persons` data source:
```

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rule-units-con.adoc:
```diff
@@ -0,0 +1,601 @@
+You can execute a rule unit in _passive mode_ by using the `run()` method (equivalent to invoking `fireAllRules()` on a KIE session)
+or in _active mode_ using the `runUntilHalt()` method (equivalent to invoking `fireUntilHalt()` on a KIE session). By default, the {DECISION_ENGINE} runs in _passive mode_ and evaluates rule units only when a user or an application explicitly calls `run()` (or `fireAllRules()` for standard rules). If a user or application calls `runUntilHalt()` for rule units (or `fireUntilHalt()` for standard ru ...
+
+If you use the `runUntilHalt()` method, invoke the method on a separate execution thread to avoid blocking the thread:
```
Done.

sterobin commented on doc-content/drools-docs/src/main/asciidoc/LanguageReference/drl-rule-units-con.adoc:
```diff
@@ -0,0 +1,601 @@
+
+A rule unit data source is a source of the data processed by a given rule unit and represents the entry point that the {DECISION_ENGINE} uses to evaluate the rule unit. A rule unit can have zero or more data sources and each `DataSource` definition declared inside a rule unit can correspond to a different entry point into the rule unit executor. Multiple rule units can share a single data source, ...
+
+A basic way to create a `DataSource` definition in a rule unit class is to use a fixed set of data, as shown in the following example:
```
Not sure. I interpreted to mean that there are many ways, but the most basic way to stick to is this one. To avoid confusion, I've just reworded to this:

```
You can create a `DataSource` definition with a fixed set of data in a rule unit class, as shown in the following example:
```

Let us know what you think, @mariofusco.

sterobin replied:
@mariofusco And your thoughts on this ^^?

mariofusco replied:
Agreed to remove that 'basic way', there isn't a way simpler or harder than another.

sterobin replied:
Ack @mariofusco . Will make the proposed change then. Thanks.

baldimir commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-event-listeners-con.adoc:
```diff
@@ -0,0 +1,59 @@
+
+= {DECISION_ENGINE_CAP} event listeners and debug logging
+
+{PRODUCT} provides a `WorkingMemoryEventManager` interface that enables you to add or remove listeners for {DECISION_ENGINE} events, such as fact insertions and rule executions. With {DECISION_ENGINE} event listeners, you can be notified of {DECISION_ENGINE} activity and separate your logging and auditing work from the core of your application.
```
I would recommend not to mention `WorkingMemoryEventManager` interface at all. It is an internal implementation detail. The listeners are added to a KieSession as is mentioned in the example in this section. This could be reworded to something like "{PRODUCT} provides an option to attach or remove listeners for {DECISION_ENGINE} events, such as fact insertions and rule executions, to a KIE session. With..."

sterobin replied:
Agreed. Changed to this:

> In {PRODUCT}, you can add or remove listeners for {DECISION_ENGINE} events, such as fact insertions and rule executions. With ...

baldimir commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-queries-con.adoc:
```diff
@@ -0,0 +1,65 @@
+.Example {DECISION_ENGINE} query
+[source,java]
+----
+QueryResults results =
```
Based on the discussion, this example is missing the DRL query definition. I found a good example in the query section in the DRL reference doc. So please consider changing this to: 

```
// Query definition from a DRL file
query "people under the age of 21"
    $person : Person( age < 21 )
end

// Using runtime API for obtaining query results
QueryResults results = ksession.getQueryResults( "people under the age of 21" );
System.out.println( "we have " + results.size() + " people under the age of 21" );

System.out.println( "These people are under the age of 21:" );

for ( QueryResultsRow row : results ) {
    Person person = ( Person ) row.get( "person" );
    System.out.println( person.getName() + "\n" );
}
```

sterobin replied:
Done

baldimir commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-queries-con.adoc:
```diff
@@ -0,0 +1,65 @@
+
+To activate a live query, start your query with parameters and monitor changes in the resulting view. You can use the `dispose()` method to terminate the query and discontinue this reactive scenario.
+
+.Example live query with a `ViewChangedEventListener` listener
```
Please add a DRL query to this example: 
```
// Query definition from a DRL file
query colors(String $color1, String $color2)
    TShirt(mainColor = $color1, secondColor = $color2, $price: manufactureCost)
end
```

sterobin replied:
Done

baldimir commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/engine-queries-con.adoc:
```diff
@@ -0,0 +1,65 @@
+[id='engine-queries-con_{context}']
+
+= {DECISION_ENGINE_CAP} queries and live queries
```
Discussed this with @sterobin - based on the discussion I added two new comments to this query section.

sterobin replied:
Fixed.

baldimir commented:
@sterobin please update also DRL guide. I found an error there:
- In section "2.4. QUERIES IN DRL"
- Example called "Example iteration over query results"
- There is currently this: 
```
QueryResults results = ksession.getQueryResults( "people over the age of 30" );
System.out.println( "we have " + results.size() + " people over the age  of 30" );
```
... please change it to this (the query name is wrong in these two lines): 
```
QueryResults results = ksession.getQueryResults( "people under the age of 21" );
System.out.println( "we have " + results.size() + " people under the age of 21" );
```

sterobin commented:
@baldimir, I've implemented your queries feedback for both the DRL and the decision engine sections, and updated the doc preview links. Can you compare the two and confirm that all is correct now? Previews below for convenience:

* [Decision engine queries and live queries](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DM/#engine-queries-con_decision-engine)
* [Queries in DRL](http://file.rdu.redhat.com/~sterobin/BXMSDOC-4468_DRL/#drl-queries-con_drl-rules)

csherrar commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/rule-execution-modes-con.adoc:
```diff
@@ -0,0 +1,107 @@
+... Thread.sleep( 1000L ); ...
+
+session.insert( tick3 );
+
```
Question - Should there be a third `... Thread.sleep( 1000L ); ...` here?

sterobin replied:
The sleep is between ticks, so the last tick doesn't require an additional sleep, because it then ends. But boy, what an eye! Thanks Clifton!

sterobin commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/cep-modes-con.adoc:
```diff
@@ -34,6 +34,20 @@ KieBaseConfiguration config = KieServices.Factory.get().newKieBaseConfiguration(
 config.setOption(EventProcessingOption.CLOUD);
 ----
+
+You can also specify cloud mode using the `eventProcessingMode="<mode>"` KIE base attribute in the KIE module descriptor file (`kmodule.xml`) for a specific {PRODUCT} project:
```
The "KIE module descriptor file" is the kmodule.xml file. They're synonymous, and comes up several times throughout the packaging doc.

mariofusco commented on doc-content/drools-docs/src/main/asciidoc/DecisionEngine/rule-execution-modes-con.adoc:
```diff
@@ -0,0 +1,107 @@
+
+The {DECISION_ENGINE} supports the following rule execution modes that determine how and when the {DECISION_ENGINE} executes rules:
+
+* *Passive mode*: (Default) The {DECISION_ENGINE} evaluates rules when a user or an application explicitly calls `fireAllRules()`. Passive mode in the {DECISION_ENGINE} is best for applications that require direct control over rule evaluation and execution, or for complex event processing (CEP) applications that use the pseudo clock implementation in the {DECISION_ENGINE}.
```
Very minor thing: in my opinion there's no default here, just 2 different methods to fire a session with different implications and the user can invoke one or the other. I'd remove the '(Default)' word there.

sterobin replied:
@mariofusco, but when you first start using our product, by default, the decision engine runs in passive mode. So I think this is accurate and important to say. Wdyt?

mariofusco replied:
@sterobin it depends on what you mean with 'default'. Normally in computer programming the default is what happens when you don't say anything, e.g.: drools works in cloud and stream mode, but if you don't say anything it runs in cloud mode, so this is the default.

In this case you cannot avoid to call one or the other fire. If you miss to do so no evaluation will happen. You have to explicitly call one of the two and in the sense there is no default behaviour. 

Conversely if with 'default' you meant 'the most commonly used choice" then this is ok for me.

sterobin replied:
@mariofusco Understood, fair point. I think it still makes sense to leave it flagged as default in the second sense, and to imply that active mode does something continually when you call it until you stop it, while passive mode does nothing until and only when you call it. Your explanation is definitely valid, I just think that not saying that one is default will leave the user wondering. So I think what is here is better than not saying anything, imo. So if you're okay with that, I'll leave it as is.
